drivers/i2c/i2c_sam0.c

```diff
@@ -0,0 +1,511 @@
+	struct k_sem sem;
+	struct i2c_sam0_msg msg;
+};
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_CFG(dev) \
+	((const struct i2c_sam0_dev_config *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2c_sam0_dev_data *const)(dev)->driver_data)
+
+static void wait_synchronization(SercomI2cm *regs)
+{
+#if defined(SERCOM_I2CM_SYNCBUSY_MASK)
+	/* SYNCBUSY is a register */
+	while ((regs->SYNCBUSY.reg & SERCOM_I2CM_SYNCBUSY_MASK) != 0) {
+	}
+#elif defined(SERCOM_I2CM_STATUS_SYNCBUSY)
+	/* SYNCBUSY is a bit */
+	while ((regs->STATUS.reg & SERCOM_I2CM_STATUS_SYNCBUSY) != 0) {
+	}
+#else
+#error Unsupported device
+#endif
+}
+
+
+static void i2c_sam0_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	struct i2c_sam0_dev_data *data = DEV_DATA(dev);
+	const struct i2c_sam0_dev_config *const cfg = DEV_CFG(dev);
+	SercomI2cm *const i2c = cfg->regs;
+
+	/* Get present interrupts and clear them */
+	u32_t status = i2c->INTFLAG.reg;
+
+	i2c->INTFLAG.reg = status;
+
+	if (i2c->STATUS.reg & (SERCOM_I2CM_STATUS_ARBLOST |
+			SERCOM_I2CM_STATUS_RXNACK |
+#ifdef SERCOM_I2CM_STATUS_LENERR
+			SERCOM_I2CM_STATUS_LENERR |
+#endif
+#ifdef SERCOM_I2CM_STATUS_SEXTTOUT
+			SERCOM_I2CM_STATUS_SEXTTOUT |
+#endif
+#ifdef SERCOM_I2CM_STATUS_MEXTTOUT
+			SERCOM_I2CM_STATUS_MEXTTOUT |
+#endif
+			SERCOM_I2CM_STATUS_LOWTOUT |
+			SERCOM_I2CM_STATUS_BUSERR)) {
+		data->msg.status = i2c->STATUS.reg;
+
+		/*
+		 * Clear all the flags that require an explicit clear
+		 * (as opposed to being cleared by ADDR writes, etc)
+		 */
+		i2c->STATUS.reg = SERCOM_I2CM_STATUS_ARBLOST |
+#ifdef SERCOM_I2CM_STATUS_LENERR
+				SERCOM_I2CM_STATUS_LENERR |
+#endif
+				SERCOM_I2CM_STATUS_LOWTOUT |
+				SERCOM_I2CM_STATUS_BUSERR;
+		wait_synchronization(i2c);
+
+		i2c->INTENCLR.reg = SERCOM_I2CM_INTENCLR_MASK;
+		k_sem_give(&data->sem);
+		return;
+	}
+
+	if (status & SERCOM_I2CM_INTFLAG_MB) {
+		if (!data->msg.size) {
+			i2c->INTENCLR.reg = SERCOM_I2CM_INTENCLR_MASK;
+			k_sem_give(&data->sem);
+			return;
+		}
+
+		i2c->DATA.reg = *data->msg.buffer;
+		data->msg.buffer++;
+		data->msg.size--;
+
+		return;
+	}
+
+	if (status & SERCOM_I2CM_INTFLAG_SB) {
+		if (data->msg.size == 1) {
+			/*
+			 * If this is the last byte, then prepare for an auto
+			 * NACK before doing the actual read, also not write
+			 * synchronized
+			 */
+			i2c->CTRLB.bit.ACKACT = 1;
+		}
+
+		*data->msg.buffer = i2c->DATA.reg;
+		data->msg.buffer++;
+		data->msg.size--;
+
+		if (!data->msg.size) {
+			i2c->INTENCLR.reg = SERCOM_I2CM_INTENCLR_MASK;
+			k_sem_give(&data->sem);
+			return;
+		}
+		return;
+	}
+}
+
+static int i2c_sam0_transfer(struct device *dev, struct i2c_msg *msgs,
+			     u8_t num_msgs, u16_t addr)
+{
+	struct i2c_sam0_dev_data *data = DEV_DATA(dev);
+	const struct i2c_sam0_dev_config *const cfg = DEV_CFG(dev);
+	SercomI2cm *const i2c = cfg->regs;
+	u32_t addr_reg;
+
+	if (!num_msgs) {
+		return 0;
+	}
+
+	for (; num_msgs > 0;) {
+		if (!msgs->len) {
```

Comment:
This breaks `samples/drivers/i2c_scanner` - just drop that test.

Reply:
It's still needed for the read case: the hardware will always do one read (see the SAMD21 manual 28.6.2.4) and my logic would underflow without it.

For the write case, you're correct and I'll drop it.  Since `samples/drivers/i2c_scanner` uses writes, that should fix it.